File: src/windowEta.ts
// get_window_eta (TRDD-8ZMZ4I6B) — how long until the CURRENT account exhausts its rate-limit
// windows, at the current COST rate. Anthropic meters the 5h/7d windows by COST (cache-read is
// weighted far below fresh input), so a token-based projection over-counts the ~96%-cache-read
// stream and under-estimates the time left. This command projects on dollars: remaining cost
// capacity ÷ the account's current $/min.
//
// Reuses accountBurners' machine-wide account-state attribution (one OAuth token active at a time)
// and its capacity resolver (own observed calibration → same-plan proxy → none) so the ETA and the
// burners autopsy can never disagree about which events belong to the account or what its cap is.

import { BILLABLE_WEIGHTS, type ConsumptionEvent, type ObservedAccountCapacity } from './burnMonitor'
import {
  eventsForAccountInWindow, resolveWindowCapacity, type AccountSegment, type ResolvedAccount, type WindowCapacity,
} from './accountBurners'

export interface WindowEtaSection {
  label: '5h' | '7d'
  windowHours: number
  window: { fromIso: string; untilIso: string }
  consumedCostUsd: number
  consumedTokens: number
  consumedBillableWeighted: number
  capacity: WindowCapacity
  /** consumedCostUsd / capacity.costUsd (0..1+), null when no cost capacity is known. */
  fillPct: number | null
  remainingCostUsd: number | null
  /** The account's current cost burn rate over `rateWindowMin`, $/min. */
  costPerMin: number
  /** Steady-state fill of a ROLLING window at the current rate: costPerMin × windowMinutes. The
   *  window sheds consumption older than its length, so if this is below the cap the window can
   *  NEVER exhaust at the current rate — it plateaus here. This is what makes a naive
   *  remaining/rate projection wrong for a rolling window. */
  steadyStateFillUsd: number
  /** True iff steadyStateFillUsd ≥ capacity — only then does an ETA exist. */
  willExhaustAtCurrentRate: boolean
  /** Minutes to exhaustion. null = no capacity, rate 0 (idle), or plateaus below cap; 0 = already over. */
  etaMinutes: number | null
  etaReason: 'projected' | 'over-limit' | 'no-capacity' | 'idle' | 'plateau'
  etaHuman: string
  exhaustionEtaIso: string | null
}

export interface WindowEtaReport {
  account: { accountId: string; email: string | null; plan: string | null; isCurrent: boolean }
  nowIso: string
  rateWindowMin: number
  fiveHour: WindowEtaSection
  sevenDay: WindowEtaSection
  /** The window that runs out FIRST at the current rate — what the user will actually hit. */
  bindingWindow: '5h' | '7d' | 'none'
  verdict: string
  note: string
  text: string
}

function weighted(e: ConsumptionEvent): number {
  const known = (e.inputTokens ?? 0) + (e.outputTokens ?? 0) + (e.cacheReadTokens ?? 0) + (e.cacheCreateTokens ?? 0)
  return (e.inputTokens ?? 0) * BILLABLE_WEIGHTS.input
    + (e.outputTokens ?? 0) * BILLABLE_WEIGHTS.output
    + (e.cacheReadTokens ?? 0) * BILLABLE_WEIGHTS.cacheRead
    + (e.cacheCreateTokens ?? 0) * BILLABLE_WEIGHTS.cacheCreate
    + Math.max(0, e.tokens - known) * BILLABLE_WEIGHTS.unknown
}

/** The readable ETA, chosen by the reason so a rolling-window plateau is never dressed up as a
 *  finite countdown. */
export function humanEta(etaMinutes: number | null, reason: WindowEtaSection['etaReason']): string {
  switch (reason) {
    case 'no-capacity': return 'unknown (no cost capacity calibrated for this account or a same-plan one)'
    case 'idle': return 'not draining (no consumption in the rate window)'
    case 'plateau': return 'won\'t exhaust at the current rate (rolling window plateaus below the cap)'
    case 'over-limit': return 'already at/over the limit'
    case 'projected': {
      const m = etaMinutes ?? 0
      const hh = Math.floor(m / 60), mm = Math.round(m % 60)
      return hh > 0 ? `${hh}h ${mm}m` : `${mm}m`
    }
  }
}

function fmtTok(n: number): string {
  if (n >= 1e9) return `${(n / 1e9).toFixed(2)}B`
  if (n >= 1e6) return `${(n / 1e6).toFixed(1)}M`
  return `${(n / 1e3).toFixed(0)}k`
}

function buildSection(opts: {
  events: ConsumptionEvent[]
  target: ResolvedAccount
  label: '5h' | '7d'
  windowHours: number
  nowMs: number
  costPerMin: number
  capacity: WindowCapacity
}): WindowEtaSection {
  const { events, target, label, windowHours, nowMs, costPerMin, capacity } = opts
  const fromMs = nowMs - windowHours * 3600_000
  const inWindow = eventsForAccountInWindow(events, target, fromMs, nowMs, nowMs)
  let cost = 0, tokens = 0, bw = 0
  for (const e of inWindow) { cost += e.costUsd; tokens += e.tokens; bw += weighted(e) }

  const cap = capacity.costUsd
  const fillPct = cap !== null && cap > 0 ? (cost / cap) * 100 : null
  const remaining = cap !== null ? cap - cost : null
  // A rolling window at steady rate r plateaus at r × windowLength. If that plateau is below the
  // cap, the window CANNOT exhaust at this rate no matter how long it runs — reporting remaining/rate
  // there would be a fiction (it assumes monotonic accumulation the rolling window never does).
  const steadyStateFillUsd = costPerMin * windowHours * 60
  const willExhaustAtCurrentRate = cap !== null && steadyStateFillUsd >= cap

  let etaReason: WindowEtaSection['etaReason']
  let etaMinutes: number | null
  if (cap === null) { etaReason = 'no-capacity'; etaMinutes = null }
  else if (remaining! <= 0) { etaReason = 'over-limit'; etaMinutes = 0 }
  else if (costPerMin <= 0) { etaReason = 'idle'; etaMinutes = null }
  else if (!willExhaustAtCurrentRate) { etaReason = 'plateau'; etaMinutes = null }
  else { etaReason = 'projected'; etaMinutes = remaining! / costPerMin }

  return {
    label, windowHours,
    window: { fromIso: new Date(fromMs).toISOString(), untilIso: new Date(nowMs).toISOString() },
    consumedCostUsd: +cost.toFixed(4),
    consumedTokens: tokens,
    consumedBillableWeighted: Math.round(bw),
    capacity,
    fillPct,
    remainingCostUsd: remaining === null ? null : +remaining.toFixed(4),
    costPerMin: +costPerMin.toFixed(6),
    steadyStateFillUsd: +steadyStateFillUsd.toFixed(2),
    willExhaustAtCurrentRate,
    etaMinutes: etaMinutes === null ? null : +etaMinutes.toFixed(1),
    etaReason,
    etaHuman: humanEta(etaMinutes, etaReason),
    exhaustionEtaIso: etaMinutes !== null && etaMinutes > 0 ? new Date(nowMs + etaMinutes * 60_000).toISOString() : null,
  }
}

export function buildWindowEtaReport(opts: {
  events: ConsumptionEvent[]
  target: ResolvedAccount
  allSegments: AccountSegment[]
  nowMs: number
  rateWindowMs: number
  observed: Record<string, ObservedAccountCapacity>
}): WindowEtaReport {
  const { events, target, allSegments, nowMs, rateWindowMs, observed } = opts

  // The account's CURRENT cost rate: its own events over the rate window (per-account, because the
  // rate limit is per account — a concurrent session on a different token does not fill this window).
  const rateFrom = nowMs - rateWindowMs
  const rateEvents = eventsForAccountInWindow(events, target, rateFrom, nowMs, nowMs)
  const rateCost = rateEvents.reduce((a, e) => a + e.costUsd, 0)
  const costPerMin = rateCost / (rateWindowMs / 60_000)

  const fiveHour = buildSection({
    events, target, label: '5h', windowHours: 5, nowMs, costPerMin,
    capacity: resolveWindowCapacity(observed, target, allSegments, '5h'),
  })
  const sevenDay = buildSection({
    events, target, label: '7d', windowHours: 168, nowMs, costPerMin,
    capacity: resolveWindowCapacity(observed, target, allSegments, '7d'),
  })

  // Which runs out first? The smaller positive ETA. Ties/None handled explicitly.
  const cand = [fiveHour, sevenDay].filter(s => s.etaMinutes !== null && s.etaMinutes > 0)
  const binding = cand.sort((a, b) => a.etaMinutes! - b.etaMinutes!)[0]
  const anyOver = [fiveHour, sevenDay].find(s => s.etaMinutes === 0)
  const bindingWindow: '5h' | '7d' | 'none' = anyOver ? anyOver.label : (binding?.label ?? 'none')

  const capSrc = (s: WindowEtaSection): string =>
    s.capacity.source === 'same-plan-proxy' ? `same-plan proxy ${s.capacity.proxyAccountId?.slice(0, 8)}` : (s.capacity.source ?? 'none')

  let verdict: string
  if (anyOver) {
    verdict = `The ${anyOver.label} window is ALREADY at/over its cost limit (${anyOver.consumedCostUsd.toFixed(0)} of ${anyOver.capacity.costUsd?.toFixed(0)} $ cap) — a rotation is imminent or overdue.`
  } else if (binding) {
    verdict = `At $${costPerMin.toFixed(2)}/min the ${binding.label} window exhausts first in ~${binding.etaHuman} ` +
      `(${binding.consumedCostUsd.toFixed(0)} of ${binding.capacity.costUsd?.toFixed(0)} $ cap, ${binding.fillPct?.toFixed(0)}% used, capacity ${capSrc(binding)}).`
  } else if (fiveHour.capacity.costUsd === null && sevenDay.capacity.costUsd === null) {
    verdict = 'No cost capacity is calibrated for this account or a same-plan account, so no ETA can be projected. A future rate-limit hit auto-calibrates it; or set AGENTLENS_WINDOW_5H_COST_USD / _7D_COST_USD.'
  } else if (costPerMin <= 0) {
    verdict = `Capacity is known but nothing is burning in the last ${Math.round(rateWindowMs / 60_000)}m (rate $0/min) — the windows are not draining right now.`
  } else {
    // Rolling-window plateau: the current rate is not enough for either window to reach its cap.
    const worst = [fiveHour, sevenDay].filter(s => s.capacity.costUsd !== null)
      .sort((a, b) => (b.steadyStateFillUsd / b.capacity.costUsd!) - (a.steadyStateFillUsd / a.capacity.costUsd!))[0]
    verdict = `At $${costPerMin.toFixed(2)}/min NEITHER window exhausts — a rolling window plateaus at rate×length. ` +
      (worst ? `The ${worst.label} plateaus at ~$${worst.steadyStateFillUsd.toFixed(0)} of its $${worst.capacity.costUsd!.toFixed(0)} cap (${Math.round(worst.steadyStateFillUsd / worst.capacity.costUsd! * 100)}%). ` : '') +
      'It would take a sustained higher burst (like the one that forced the last rotation) to exhaust a window.'
  }

  const note = 'ETA is COST-based (Anthropic meters the windows by cost, not raw tokens — cache-read is weighted ~0.1×). ' +
    'The rate is THIS account\'s own $/min (rate limits are per OAuth account); capacity is the account\'s observed ' +
    'calibration, else a same-plan account\'s as a labeled proxy, else undetermined. Consumption uses time-based ' +
    'attribution against the account-state timeline, so cross-rotation sessions split correctly.'

  const lines: string[] = []
  const acct = `${target.email ?? target.accountId}${target.isCurrent ? '' : ' (NOT the current account)'}`
  lines.push(`window ETA for ${acct} at $${costPerMin.toFixed(2)}/min (rate over last ${Math.round(rateWindowMs / 60_000)}m) — as of ${new Date(nowMs).toISOString()}`)
  for (const s of [fiveHour, sevenDay]) {
    const mark = bindingWindow === s.label ? '  ◀ EXHAUSTS FIRST' : ''
    const capStr = s.capacity.costUsd !== null
      ? `$${s.consumedCostUsd.toFixed(0)} / $${s.capacity.costUsd.toFixed(0)} cap (${s.fillPct?.toFixed(0)}%, ${capSrc(s)})`
      : `$${s.consumedCostUsd.toFixed(0)} consumed · no capacity`
    lines.push(`  ${s.label}: ${capStr} · ${fmtTok(s.consumedTokens)} raw tok · ETA ${s.etaHuman}${s.exhaustionEtaIso ? ` (≈ ${s.exhaustionEtaIso})` : ''}${mark}`)
  }
  lines.push(verdict)

  return {
    account: { accountId: target.accountId, email: target.email, plan: target.plan, isCurrent: target.isCurrent },
    nowIso: new Date(nowMs).toISOString(),
    rateWindowMin: Math.round(rateWindowMs / 60_000),
    fiveHour, sevenDay,
    bindingWindow,
    verdict,
    note,
    text: lines.join('\n'),
  }
}

File: rust-core/crates/agentlens-core/src/window_eta.rs
//! Port of src/windowEta.ts (TRDD-8ZMZ4I6B, ported under TRDD-DMWOBWFH P4x.2d) —
//! `get_window_eta`: how long until the CURRENT account exhausts its rate-limit windows, at the
//! current COST rate.
//!
//! Anthropic meters the 5h/7d windows by COST (cache-read is weighted far below fresh input), so a
//! token-based projection over-counts the ~96%-cache-read stream and under-estimates the time left.
//! This projects on dollars: remaining cost capacity ÷ the account's current $/min.
//!
//! Reuses `account_burners`' machine-wide attribution and its capacity resolver (own observed
//! calibration → same-plan proxy → none) so the ETA and the burners autopsy can never disagree
//! about which events belong to the account or what its cap is.

use indexmap::IndexMap;
use serde_json::{Map, Value};

use crate::account_burners::{
    events_for_account_in_window, fmt_tok, resolve_window_capacity, weighted, AccountSegment, ResolvedAccount, WindowCapacity,
};
use crate::burn::monitor::ObservedAccountCapacity;
use crate::summarize::helpers::{iso_from_ms, js_math_round, js_to_fixed_num, js_to_fixed_str, num};

/// The readable ETA, chosen by the REASON so a rolling-window plateau is never dressed up as a
/// finite countdown — the whole point of separating the two.
pub fn human_eta(eta_minutes: Option<f64>, reason: &str) -> String {
    match reason {
        "no-capacity" => "unknown (no cost capacity calibrated for this account or a same-plan one)".to_owned(),
        "idle" => "not draining (no consumption in the rate window)".to_owned(),
        "plateau" => "won't exhaust at the current rate (rolling window plateaus below the cap)".to_owned(),
        "over-limit" => "already at/over the limit".to_owned(),
        _ => {
            let m = eta_minutes.unwrap_or(0.0);
            let hh = (m / 60.0).floor();
            // `Math.round`, which is half toward +∞ — NOT Rust's half-away-from-zero.
            let mm = js_math_round(m % 60.0);
            if hh > 0.0 {
                format!("{}h {}m", crate::summarize::helpers::fmt_js_num(hh), crate::summarize::helpers::fmt_js_num(mm))
            } else {
                format!("{}m", crate::summarize::helpers::fmt_js_num(mm))
            }
        }
    }
}

pub struct EtaSection {
    label: &'static str,
    window_hours: f64,
    from_ms: f64,
    now_ms: f64,
    consumed_cost_usd: f64,
    consumed_tokens: f64,
    consumed_billable_weighted: f64,
    capacity: WindowCapacity,
    fill_pct: Option<f64>,
    remaining_cost_usd: Option<f64>,
    cost_per_min: f64,
    steady_state_fill_usd: f64,
    will_exhaust: bool,
    /// The ROUNDED field (`+etaMinutes.toFixed(1)`) — what the report reports and what the binding
    /// pick compares.
    eta_minutes: Option<f64>,
    eta_reason: &'static str,
    eta_human: String,
    exhaustion_eta_iso: Option<String>,
}

impl EtaSection {
    fn to_value(&self) -> Value {
        let mut w = Map::new();
        w.insert("fromIso".into(), Value::String(iso_from_ms(self.from_ms)));
        w.insert("untilIso".into(), Value::String(iso_from_ms(self.now_ms)));
        let mut m = Map::new();
        m.insert("label".into(), Value::String(self.label.to_owned()));
        m.insert("windowHours".into(), num(self.window_hours));
        m.insert("window".into(), Value::Object(w));
        m.insert("consumedCostUsd".into(), num(self.consumed_cost_usd));
        m.insert("consumedTokens".into(), num(self.consumed_tokens));
        m.insert("consumedBillableWeighted".into(), num(self.consumed_billable_weighted));
        m.insert("capacity".into(), self.capacity.to_value());
        m.insert("fillPct".into(), self.fill_pct.map(num).unwrap_or(Value::Null));
        m.insert("remainingCostUsd".into(), self.remaining_cost_usd.map(num).unwrap_or(Value::Null));
        m.insert("costPerMin".into(), num(self.cost_per_min));
        m.insert("steadyStateFillUsd".into(), num(self.steady_state_fill_usd));
        m.insert("willExhaustAtCurrentRate".into(), Value::Bool(self.will_exhaust));
        m.insert("etaMinutes".into(), self.eta_minutes.map(num).unwrap_or(Value::Null));
        m.insert("etaReason".into(), Value::String(self.eta_reason.to_owned()));
        m.insert("etaHuman".into(), Value::String(self.eta_human.clone()));
        m.insert("exhaustionEtaIso".into(), self.exhaustion_eta_iso.clone().map(Value::from).unwrap_or(Value::Null));
        Value::Object(m)
    }
}

struct SectionOpts<'a> {
    events: &'a [Value],
    target: &'a ResolvedAccount,
    label: &'static str,
    window_hours: f64,
    now_ms: f64,
    cost_per_min: f64,
    capacity: WindowCapacity,
}

fn build_section(o: SectionOpts<'_>) -> EtaSection {
    let from_ms = o.now_ms - o.window_hours * 3_600_000.0;
    let in_window = events_for_account_in_window(o.events, o.target, from_ms, o.now_ms, o.now_ms);
    let (mut cost, mut tokens, mut bw) = (0.0_f64, 0.0_f64, 0.0_f64);
    for e in &in_window {
        cost += e.get("costUsd").and_then(Value::as_f64).unwrap_or(0.0);
        tokens += e.get("tokens").and_then(Value::as_f64).unwrap_or(0.0);
        bw += weighted(e);
    }

    let cap = o.capacity.cost_usd;
    let fill_pct = cap.filter(|c| *c > 0.0).map(|c| (cost / c) * 100.0);
    let remaining = cap.map(|c| c - cost);
    // A rolling window at steady rate r plateaus at r × windowLength. If that plateau is below the
    // cap the window CANNOT exhaust at this rate however long it runs — reporting remaining/rate
    // there would be a fiction, because it assumes a monotonic accumulation a rolling window never
    // does. This is the distinction a naive projection gets wrong.
    let steady_state_fill_usd = o.cost_per_min * o.window_hours * 60.0;
    let will_exhaust = cap.is_some_and(|c| steady_state_fill_usd >= c);

    let (eta_reason, eta_minutes_raw): (&'static str, Option<f64>) = if cap.is_none() {
        ("no-capacity", None)
    } else if remaining.unwrap() <= 0.0 {
        ("over-limit", Some(0.0))
    } else if o.cost_per_min <= 0.0 {
        ("idle", None)
    } else if !will_exhaust {
        ("plateau", None)
    } else {
        ("projected", Some(remaining.unwrap() / o.cost_per_min))
    };

    EtaSection {
        label: o.label,
        window_hours: o.window_hours,
        from_ms,
        now_ms: o.now_ms,
        consumed_cost_usd: js_to_fixed_num(cost, 4),
        consumed_tokens: tokens,
        consumed_billable_weighted: js_math_round(bw),
        capacity: o.capacity,
        fill_pct,
        remaining_cost_usd: remaining.map(|r| js_to_fixed_num(r, 4)),
        cost_per_min: js_to_fixed_num(o.cost_per_min, 6),
        steady_state_fill_usd: js_to_fixed_num(steady_state_fill_usd, 2),
        will_exhaust,
        eta_minutes: eta_minutes_raw.map(|m| js_to_fixed_num(m, 1)),
        eta_reason,
        // `humanEta(etaMinutes, …)` is called with the UNROUNDED local, not the rounded field.
        eta_human: human_eta(eta_minutes_raw, eta_reason),
        // …and so is the exhaustion instant: `new Date(nowMs + etaMinutes * 60_000)` reads the
        // UNROUNDED value too, so the ISO can disagree with `etaMinutes × 60s` by up to 3 seconds.
        // Rounding first would silently "fix" a discrepancy the TS actually emits.
        exhaustion_eta_iso: eta_minutes_raw.filter(|m| *m > 0.0).map(|m| iso_from_ms(o.now_ms + m * 60_000.0)),
    }
}

fn cap_src(s: &EtaSection) -> String {
    match s.capacity.source {
        Some("same-plan-proxy") => {
            let id = s.capacity.proxy_account_id.as_deref().map_or("undefined".to_owned(), |p| p.chars().take(8).collect());
            format!("same-plan proxy {id}")
        }
        Some(other) => other.to_owned(),
        None => "none".to_owned(),
    }
}

/// `x?.toFixed(n)` inside a template literal — an absent value stringifies to "undefined", which is
/// what the TS actually renders. Silently substituting "0" or "" would print a different report.
fn opt_fixed(x: Option<f64>, digits: usize) -> String {
    x.map_or("undefined".to_owned(), |v| js_to_fixed_str(v, digits))
}

pub struct WindowEtaOpts<'a> {
    pub events: &'a [Value],
    pub target: &'a ResolvedAccount,
    pub all_segments: &'a [AccountSegment],
    pub now_ms: f64,
    pub rate_window_ms: f64,
    pub observed: &'a IndexMap<String, ObservedAccountCapacity>,
}

pub fn build_window_eta_report(o: &WindowEtaOpts<'_>) -> Value {
    // The account's CURRENT cost rate: its OWN events over the rate window — per-account, because
    // the rate limit is per account and a concurrent session on a different token does not fill it.
    let rate_from = o.now_ms - o.rate_window_ms;
    let rate_events = events_for_account_in_window(o.events, o.target, rate_from, o.now_ms, o.now_ms);
    let rate_cost: f64 = rate_events.iter().map(|e| e.get("costUsd").and_then(Value::as_f64).unwrap_or(0.0)).sum();
    let cost_per_min = rate_cost / (o.rate_window_ms / 60_000.0);

    let five_hour = build_section(SectionOpts {
        events: o.events,
        target: o.target,
        label: "5h",
        window_hours: 5.0,
        now_ms: o.now_ms,
        cost_per_min,
        capacity: resolve_window_capacity(o.observed, o.target, o.all_segments, "5h"),
    });
    let seven_day = build_section(SectionOpts {
        events: o.events,
        target: o.target,
        label: "7d",
        window_hours: 168.0,
        now_ms: o.now_ms,
        cost_per_min,
        capacity: resolve_window_capacity(o.observed, o.target, o.all_segments, "7d"),
    });

    let sections = [&five_hour, &seven_day];
    // Which runs out FIRST? The smaller positive ETA — but an already-over window wins outright,
    // because "you are past the limit" is not a countdown.
    let mut cand: Vec<&EtaSection> = sections.iter().copied().filter(|s| s.eta_minutes.is_some_and(|m| m > 0.0)).collect();
    cand.sort_by(|a, b| (a.eta_minutes.unwrap() - b.eta_minutes.unwrap()).partial_cmp(&0.0).unwrap_or(std::cmp::Ordering::Equal));
    let binding = cand.first().copied();
    let any_over = sections.iter().copied().find(|s| s.eta_minutes == Some(0.0));
    let binding_window = any_over.map(|s| s.label).or_else(|| binding.map(|s| s.label)).unwrap_or("none");

    let rate_window_min = js_math_round(o.rate_window_ms / 60_000.0);
    let verdict = if let Some(over) = any_over {
        format!(
            "The {} window is ALREADY at/over its cost limit ({} of {} $ cap) — a rotation is imminent or overdue.",
            over.label,
            js_to_fixed_str(over.consumed_cost_usd, 0),
            opt_fixed(over.capacity.cost_usd, 0)
        )
    } else if let Some(b) = binding {
        format!(
            "At ${}/min the {} window exhausts first in ~{} ({} of {} $ cap, {}% used, capacity {}).",
            js_to_fixed_str(cost_per_min, 2),
            b.label,
            b.eta_human,
            js_to_fixed_str(b.consumed_cost_usd, 0),
            opt_fixed(b.capacity.cost_usd, 0),
            opt_fixed(b.fill_pct, 0),
            cap_src(b)
        )
    } else if five_hour.capacity.cost_usd.is_none() && seven_day.capacity.cost_usd.is_none() {
        "No cost capacity is calibrated for this account or a same-plan account, so no ETA can be projected. A future rate-limit hit auto-calibrates it; or set AGENTLENS_WINDOW_5H_COST_USD / _7D_COST_USD.".to_owned()
    } else if cost_per_min <= 0.0 {
        format!(
            "Capacity is known but nothing is burning in the last {}m (rate $0/min) — the windows are not draining right now.",
            crate::summarize::helpers::fmt_js_num(rate_window_min)
        )
    } else {
        // Rolling-window plateau: the current rate is not enough for EITHER window to reach its cap.
        let mut with_cap: Vec<&EtaSection> = sections.iter().copied().filter(|s| s.capacity.cost_usd.is_some()).collect();
        with_cap.sort_by(|a, b| {
            let r = |s: &EtaSection| s.steady_state_fill_usd / s.capacity.cost_usd.unwrap();
            (r(b) - r(a)).partial_cmp(&0.0).unwrap_or(std::cmp::Ordering::Equal)
        });
        let worst_txt = with_cap.first().map_or(String::new(), |w| {
            let cap = w.capacity.cost_usd.unwrap();
            format!(
                "The {} plateaus at ~${} of its ${} cap ({}%). ",
                w.label,
                js_to_fixed_str(w.steady_state_fill_usd, 0),
                js_to_fixed_str(cap, 0),
                crate::summarize::helpers::fmt_js_num(js_math_round(w.steady_state_fill_usd / cap * 100.0))
            )
        });
        format!(
            "At ${}/min NEITHER window exhausts — a rolling window plateaus at rate×length. {}It would take a sustained higher burst (like the one that forced the last rotation) to exhaust a window.",
            js_to_fixed_str(cost_per_min, 2),
            worst_txt
        )
    };

    let note = "ETA is COST-based (Anthropic meters the windows by cost, not raw tokens — cache-read is weighted ~0.1×). The rate is THIS account's own $/min (rate limits are per OAuth account); capacity is the account's observed calibration, else a same-plan account's as a labeled proxy, else undetermined. Consumption uses time-based attribution against the account-state timeline, so cross-rotation sessions split correctly.";

    let acct = format!(
        "{}{}",
        o.target.email.clone().unwrap_or_else(|| o.target.account_id.clone()),
        if o.target.is_current { "" } else { " (NOT the current account)" }
    );
    let mut lines = vec![format!(
        "window ETA for {} at ${}/min (rate over last {}m) — as of {}",
        acct,
        js_to_fixed_str(cost_per_min, 2),
        crate::summarize::helpers::fmt_js_num(rate_window_min),
        iso_from_ms(o.now_ms)
    )];
    for s in sections {
        let mark = if binding_window == s.label { "  ◀ EXHAUSTS FIRST" } else { "" };
        let cap_str = match s.capacity.cost_usd {
            Some(c) => format!(
                "${} / ${} cap ({}%, {})",
                js_to_fixed_str(s.consumed_cost_usd, 0),
                js_to_fixed_str(c, 0),
                opt_fixed(s.fill_pct, 0),
                cap_src(s)
            ),
            None => format!("${} consumed · no capacity", js_to_fixed_str(s.consumed_cost_usd, 0)),
        };
        lines.push(format!(
            "  {}: {} · {} raw tok · ETA {}{}{}",
            s.label,
            cap_str,
            fmt_tok(s.consumed_tokens),
            s.eta_human,
            s.exhaustion_eta_iso.as_deref().map_or(String::new(), |iso| format!(" (≈ {iso})")),
            mark
        ));
    }
    lines.push(verdict.clone());

    let mut account = Map::new();
    account.insert("accountId".into(), Value::String(o.target.account_id.clone()));
    account.insert("email".into(), o.target.email.clone().map(Value::from).unwrap_or(Value::Null));
    account.insert("plan".into(), o.target.plan.clone().map(Value::from).unwrap_or(Value::Null));
    account.insert("isCurrent".into(), Value::Bool(o.target.is_current));

    let mut m = Map::new();
    m.insert("account".into(), Value::Object(account));
    m.insert("nowIso".into(), Value::String(iso_from_ms(o.now_ms)));
    m.insert("rateWindowMin".into(), num(rate_window_min));
    m.insert("fiveHour".into(), five_hour.to_value());
    m.insert("sevenDay".into(), seven_day.to_value());
    m.insert("bindingWindow".into(), Value::String(binding_window.to_owned()));
    m.insert("verdict".into(), Value::String(verdict));
    m.insert("note".into(), Value::String(note.to_owned()));
    m.insert("text".into(), Value::String(lines.join("\n")));
    Value::Object(m)
}
